Shake/Main.lean

```diff
@@ -1,39 +1,130 @@
+/-
+Copyright (c) 2023 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro
+-/
 import Lean.Util.FoldConsts
 import Lean
 
+/-! # `lake exe shake` command
+
+This command will check mathlib (or the specified modules) and all their dependency for unused
+imports. This works by looking at generated `.olean` files to deduce required imports and ensuring
+that every import is used to contribute some constant. Because recompilation is not needed this
+is quite fast (about 8 seconds to check `Mathlib` and all dependencies), but it has some known
+limitations:
+
+* Tactics that are used during elaboration generally leave no trace in the proof term, so
+  they will be incorrectly marked as unused.
+* Similarly, files that contribute only notations will not be detected.
+* Conversely, files that define tactics and notations are also likely to have false positives
+  because the notation itself does not depend on the referenced constant (it elaborates to a
+  name literal rather than an actual reference to the target constant).
+
+To mitigate this, the `scripts/noshake.json` file is used to suppress known false positives. See
+`ShakeCfg` for information regarding the file format.
+
+-/
+
+def help : String := "Mathlib4 tree shaking tool
+Usage: lake exe shake [OPTIONS] <MODULE>..
+
+Arguments:
+  <MODULE>
+    A module path like `Mathlib`. All files transitively reachable from the
+    provided module(s) will be checked.
+
+Options:
+  --fix
+    Apply the suggested fixes directly. Make sure you have a clean checkout
+    before running this, so you can review the changes.
+
+  --cfg <FILE>   (default: scripts/noshake.json)
+    Use FILE to specify which imports we should ignore.
+
+  --update
+    Assume that all issues we find are false positives and update the config
+    file to include them.
+
+  --no-downstream
+    Unless disabled, shake will check downstream files that were transitively
+    depending on the import we want to remove and re-add the import to these
+    downstream files.
+
+# The noshake.json file
+
+The file passed in the --cfg argument is a JSON file with the following
+structure:
+
+  {
+    \"ignoreAll\": [NAME],
+    \"ignoreImport\": [NAME],
+    \"ignore\": {NAME: [NAME]}
+  }
+
+The fields can be omitted if empty. They have the following interpretation:
+
+* ignoreAll:
+  All imports in these files should be treated as necessary
+* ignore[X]:
+  All imports in the list should be treated as necessary when processing X
+* ignoreImport:
+  These files should be treated as necessary when imported into any other file.
+"
+
 open Lean
 
+/-- We use `Nat` as a bitset for doing efficient set operations.
+The bit indexes will usually be a module index. -/
 abbrev Bitset := Nat
```

Comment:
Would this be reasonable in `Std` as a one-field `structure` with a `Membership` instance? It would be nice to hide the bit operators from the API, and instead have it appear as a set.

Reply:
(Maybe avoiding a dependency on Std is a feature here...)

Reply:
I don't really feel like designing this API, and in any case it's independent of this PR.